### PR TITLE
WIP: SSS in python

### PR DESCRIPTION
**Pull request by kambysese**
@Eric89GXL added support for maxwell filtering with mne-python and tested with example data. Made a number of changes I'd like to hear if you're ok with or not. LMKWYT at your earliest convenience. cc @mdclarke

### CODE REVIEW

**Comment by kambysese** [comment]
@Eric89GXL please review. I will make new PR for average movecomp.

File: examples/funloc/analysis_fun.py [hunk 1]
```diff
@@ -40,7 +40,7 @@
 
 params = mnefun.Params(tmin=-0.2, tmax=0.5, t_adjust=-4e-3,
                        n_jobs=6, n_jobs_mkl=1,
-                       n_jobs_fir='cuda', n_jobs_resample='cuda',
+                       n_jobs_fir=6, n_jobs_resample=6,
                        decim=5, proj_sfreq=200, filter_length='5s')
```
Comment (kambysese) [inline examples/funloc/analysis_fun.py]: will revert here.

File: examples/funloc/analysis_fun.py [hunk 2]
```diff
@@ -49,10 +49,10 @@
 params.subject_indices = np.arange(2)  # which subjects to run
 params.plot_drop_logs = False  # turn off for demo or plots will block
 
-params.acq_ssh = 'minea'  # can also be e.g., "eric@minea.ilabs.uw.edu"
+params.acq_ssh = "kambiz@minea.ilabs.uw.edu"
 params.acq_dir = '/sinuhe/data02/eric_non_space'
-params.sws_ssh = 'kasga'
-params.sws_dir = '/data06/larsoner'
+params.sws_ssh = 'kam@kasga.ilabs.uw.edu'
+params.sws_dir = '/data01/kam'
```
Comment (kambysese) [inline examples/funloc/analysis_fun.py]: will revert here.

File: mnefun/_mnefun.py [hunk 3]
```diff
@@ -213,7 +233,9 @@ def __init__(self, tmin=None, tmax=None, t_adjust=0, bmin=-0.2, bmax=0.0,
                  cov_method='empirical', ssp_eog_reject=None,
                  ssp_ecg_reject=None, baseline='individual',
                  reject_tmin=None, reject_tmax=None,
-                 lp_trans=0.5, hp_trans=0.5, movecomp='inter'):
+                 lp_trans=0.5, hp_trans=0.5, movecomp='inter',
+                 sss_type='maxfilter', int_order=3, ext_order=6, st_duration=10.0,
+                 st_correlation=0.98):
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: I don't think these defaults don't match what we've been using for the MaxFiltered version of the data, can you check the run_sss.sh script? It's probably worth noting in the docs above that some of these params are only used with the Python version of the algorithm.

File: mnefun/_mnefun.py [hunk 4]
```diff
@@ -842,6 +872,42 @@ def run_sss_positions(fname_in, fname_out, host='kasga', opts='', port=22):
     run_subprocess(cmd)
 
 
+def run_sss_localy(p, subjects, run_indices):
+
+    cal_file = op.join(op.dirname(op.dirname(__file__)), 'sss_cal.dat')
+    ct_file = op.join(op.dirname(op.dirname(__file__)), 'ct_sparse.fif')
+    if p.st_duration is not None:
+        st_duration = float(p.st_duration)
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: I don't think you need to do this -- mne-python should convert it to float if it's not None already.

**Comment by Eric89GXL** [comment]
Otherwise looks like a good start to me

**Comment by kambysese** [comment]
@Eric89GXL please review additional modifications. I tried testing on my linux machine and ran into 

>AttributeError: 'sss_type' is not an attribute of class <mnefun._mnefun.Params object at 0x7f9e61749350>. Call "unfreeze()" to allow addition of new attributes'

Not sure how to 'unfreeze' a class & why I am suddenly running into this?

**Comment by kambysese** [comment]
@Eric89GXL strangely I don't encounter the above `AttributeError` testing this PR on my Mac OSx installation of  Anaconda?

**Comment by kambysese** [comment]
@Eric89GXL testing on example data with most recent additions throws 

>    Computing regularization
/Users/ktavabi/Projects/mne-python/mne/preprocessing/maxwell.py:1526: RuntimeWarning: invalid value encountered in double_scalars
  % (I_tots[lim_idx], 100 * I_tots[lim_idx] / max_info,
        Resulting information: 0.0 bits/sample (nan% of peak 0.0)
        Using 80/80 inside and 15/15 outside harmonic components
Traceback (most recent call last):
  File "/Users/ktavabi/Projects/mnefun/examples/funloc/analysis_fun.py", line 114, in <module>
    print_status=False,  # Print completeness status update
  File "/Users/ktavabi/Projects/mnefun/mnefun/_mnefun.py", line 570, in do_processing
    outs[ii] = func(p, subjects, run_indices)
  File "/Users/ktavabi/Projects/mnefun/mnefun/_mnefun.py", line 739, in run_sss
    run_sss_localy(p, subjects, run_indices)
  File "/Users/ktavabi/Projects/mnefun/mnefun/_mnefun.py", line 975, in run_sss_localy
    destination=trans_to)
  File "<string>", line 2, in maxwell_filter
  File "/Users/ktavabi/Projects/mne-python/mne/utils.py", line 551, in verbose
    return function(*args, **kwargs)
  File "/Users/ktavabi/Projects/mne-python/mne/preprocessing/maxwell.py", line 341, in maxwell_filter
    raise RuntimeError(msg)
RuntimeError: Matrix is badly conditioned: 22500733164779524 >= 1000

Any ideas what is happening?

**Comment by Eric89GXL** [comment]
Information of 0.0 is really weird. Is it easy to share the file and command?

**Comment by kambysese** [comment]
I am running into this testing with analysis_fun and subj_01 of funloc the example datasets. The command in question is:

>raw_sss = maxwell_filter(raw, origin=origin, int_order=8, ext_order=3,
                                     calibration=cal_file,
                                     cross_talk=ct_file,
                                     st_correlation=0.98, st_duration=60.,
                                     destination=trans_to)

where `trans_to` is the path to median hp file created using `calc_median_hp`. NB I encountered this after adding kwargs for `destination` and using 8 and 3 for SSS bases.

**Comment by kambysese** [comment]
BTW do you have any idea why 
>AttributeError: 'sss_type' is not an attribute of class <mnefun._mnefun.Params object at 0x7fa292e3e350>. Call "unfreeze()" to allow addition of new attributes

happens on my linux machine and not my OSx? which should have similar if not identical python environments and IDE.

**Comment by Eric89GXL** [comment]
It should happen on both. You should set the `sss_type` before the freezing function is called

File: examples/funloc/analysis_fun.py [hunk 5]
```diff
@@ -66,7 +66,7 @@
                     [1, 1, 2],  # EOG
                     [0, 0, 0]]  # Continuous (from ERM)
 params.cov_method = 'shrunk'  # cleaner noise covariance regularization
-
+params.sss_type = 'python'  # python | maxfilter for choosing SSS applied using either Maxfilter or mne-python
```
Comment (Eric89GXL) [inline examples/funloc/analysis_fun.py]: this shouldn't break anything because the `sss_type` parameter should already exist

**Comment by Eric89GXL** [comment]
Are you running different versions of python on the two machines (e.g., 2.7 vs 3.4)?

**Comment by kambysese** [comment]
No. I am running 2.7.11 on anaconda 2.4.1 on both machines. It's annoying. I am stepping through right now to see when/why `params.frozen=True` prior to declaring that attribute.

**Comment by kambysese** [comment]
@Eric89GXL Jesus H Christ. I hadn't installed mnefun as a developer on the linux box. So it was running master whilst I was passing attributes that didn't exist. :disappointed: :gun: Go ahead laugh. Now it's just this "badly conditional matrix RuntimeError"

**Comment by kambysese** [comment]
So I don't know where you are with this one, these are the commands I used to reproduce the error:

>raw = mnefun._mnefun._raw_LRFCP(['subj_01_funloc_raw.fif'], 200, None, 55., 1, 'cuda',[], 'subj_01_prebad.txt', allow_maxshield=True, force_bads=True, apply_proj=False)

>raw_sss = mne.preprocessing.maxwell_filter(raw, origin=origin, int_order=3, ext_order=8,
                                     calibration='sss_cal.dat',
                                     cross_talk='ct_sparse.fif',
                                     st_correlation=.98, st_duration=60., destination='subj_01_median_pos.fif')

Do you still want these files?

**Comment by Eric89GXL** [comment]
I'll try to reproduce locally with the source files

**Comment by Eric89GXL** [comment]
What is the "origin" argument you are using?

**Comment by Eric89GXL** [comment]
also, int_order=3 ext_order=8 is reversed from how it normally is done...

**Comment by Eric89GXL** [comment]
Using `origin='auto'` and the (incorrect) `int_order=3, ext_order=8` I get:
```
    Computing regularization
        Resulting information: 77.2 bits/sample (100.0% of peak 77.2)
        Using 15/15 inside and 80/80 outside harmonic components
Traceback (most recent call last):
  File "/home/larsoner/Desktop/untitled0.py", line 15, in <module>
    st_correlation=.98, st_duration=60., destination='subj_01_median_pos.fif', verbose=True)
  File "<string>", line 2, in maxwell_filter
  File "/home/larsoner/custombuilds/mne-python/mne/utils.py", line 547, in verbose
    return function(*args, **kwargs)
  File "/home/larsoner/custombuilds/mne-python/mne/preprocessing/maxwell.py", line 360, in maxwell_filter
    info['dev_head_t'])
  File "<string>", line 2, in _get_decomp
  File "/home/larsoner/custombuilds/mne-python/mne/utils.py", line 548, in verbose
    return function(*args, **kwargs)
  File "/home/larsoner/custombuilds/mne-python/mne/preprocessing/maxwell.py", line 686, in _get_decomp
    raise RuntimeError(msg)
RuntimeError: Matrix is badly conditioned: 1103 >= 1000
```
But then switching to `int_order=8, ext_order=3` it works fine:
```
    Computing regularization
        Resulting information: 382.2 bits/sample (98.7% of peak 387.3)
        Using 72/80 inside and 15/15 outside harmonic components
    Processing 5 data chunks of (at least) 60.0 sec
        Projecting  3 intersecting tSSS components for     0.00 -    59.99 sec
        Used  1 head position for     0.00 -    59.99 sec
        Projecting  3 intersecting tSSS components for    60.00 -   120.00 sec
        Used  1 head position for    60.00 -   120.00 sec
        Projecting  4 intersecting tSSS components for   120.00 -   180.00 sec
        Used  1 head position for   120.00 -   180.00 sec
        Projecting  2 intersecting tSSS components for   180.00 -   240.00 sec
        Used  1 head position for   180.00 -   240.00 sec
        Projecting  3 intersecting tSSS components for   240.00 -   344.00 sec
        Used  1 head position for   240.00 -   344.00 sec
[done]
```
I suspect, then, that your `origin` argument is wrong. Maybe it's in mm instead of m?

**Comment by kambysese** [comment]
I mixed up `int_order` and `ext_order` coming off of infant data; that is fixed. It looks like origin was being returned by `fit_sphere_to_headshape` in mm and I was under the impression that `maxwell_filter' requires it in mm but apparently not. It looks like that was the culprit. Thanks.

**Comment by kambysese** [comment]
@Eric89GXL I am thinking it's ready for merge. Please review. Set default SSS option to use mne-python in example analysis script; so that hopefully more users kick the tires.

File: examples/funloc/analysis_fun.py [hunk 6]
```diff
@@ -96,9 +96,9 @@
     do_score=True,  # do scoring
     # Make SUBJ/raw_fif/SUBJ_prebad.txt file with space-separated
     # list of bad MEG channel numbers, needed for running SSS.
-    push_raw=True,  # Push raw files and SSS script to SSS workstation
-    do_sss=True,  # Run SSS remotely
-    fetch_sss=True,  # Fetch SSSed files
+    push_raw=False,  # Push raw files and SSS script to SSS workstation
+    do_sss=True,  # Run SSS remote using Maxfilter or local with mne-python
+    fetch_sss=False,  # Fetch SSSed files from SSS workstation
```
Comment (Eric89GXL) [inline examples/funloc/analysis_fun.py]: I think if these are `True` and `sss_type == 'python'` they should just do nothing (return without pushing or fetching), so these changes can be reverted once those changes are made.

File: mnefun/_mnefun.py [hunk 7]
```diff
@@ -186,6 +187,38 @@ class Params(Frozen):
...
+        containing a MEG device<->head transformation, or a 3-element array
+        giving the coordinates to translate to (with no rotations).
+        For example, ``destination=(0, 0, 0.04)`` would translate the bases
+        as ``--trans default`` would in MaxFilter™ (i.e., to the default
+        head location).
+    sss_origin : array-like, shape (3,) | str
+        Origin of internal and external multipolar moment space in meters. Default is
+        center of sphere fit to digitized head points.
+    coord_frame : str
+        The coordinate frame that the ``sss_origin`` is specified in, default is 'head'.
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: you shouldn't need to specify `coord_frame`, `mnefun` should do it automatically ('head' for subject files, 'device' for empty-room files)

File: mnefun/_mnefun.py [hunk 8]
```diff
@@ -186,6 +187,38 @@ class Params(Frozen):
...
+        The destination location for the head. Can be ``None``, which
+        will not change the head position, or a string path to a FIF file
+        containing a MEG device<->head transformation, or a 3-element array
+        giving the coordinates to translate to (with no rotations).
+        For example, ``destination=(0, 0, 0.04)`` would translate the bases
+        as ``--trans default`` would in MaxFilter™ (i.e., to the default
+        head location).
+    sss_origin : array-like, shape (3,) | str
+        Origin of internal and external multipolar moment space in meters. Default is
+        center of sphere fit to digitized head points.
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: I don't think for now we should expose all these to the user. Can't we just use what we used in the MaxFilter case (8/3, 60., 0.98, translate to median pos, origin auto-fit)?

File: mnefun/_mnefun.py [hunk 9]
```diff
@@ -186,6 +187,38 @@ class Params(Frozen):
...
+        Correlation limit between inner and outer subspaces used to reject
+        ovwrlapping intersecting inner/outer signals during spatiotemporal SSS.
+    trans_to : str | array-like, shape (3,) | None
+        The destination location for the head. Can be ``None``, which
+        will not change the head position, or a string path to a FIF file
+        containing a MEG device<->head transformation, or a 3-element array
+        giving the coordinates to translate to (with no rotations).
+        For example, ``destination=(0, 0, 0.04)`` would translate the bases
+        as ``--trans default`` would in MaxFilter™ (i.e., to the default
+        head location).
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: I don't think this one is actually correct --  we only allow `'median'` and `'default'`, right? Whatever we allow, we should allow the same set of options for the `mne-python` version and the `MaxFilter` verison, and they should behave the same way.

File: mnefun/_mnefun.py [hunk 10]
```diff
@@ -842,6 +906,81 @@ def run_sss_positions(fname_in, fname_out, host='kasga', opts='', port=22):
...
+
+    Notes
+    -----
+    Compared to Elekta's MaxFilter™ software, Maxwwell filtering in mne-python
+    does not require/support the following arguments:
+        -format
+        -hpicons
+        -autobad
+        -force
+        -movecomp
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: movement compensation is supported on the `dev` release

Comment (Eric89GXL) [inline mnefun/_mnefun.py]: I mean `master`, rather

File: sss_cal.dat [hunk 11]
```diff
@@ -0,0 +1,306 @@
+113	-0.106600 0.046400 -0.060400 -0.012700 0.005700 -0.999903 -0.186801 -0.982403 -0.003300 -0.981487 0.190546 0.016294 -0.000907
```
Comment (Eric89GXL) [inline sss_cal.dat]: this file and the `cal` file should both go in the `mnefun/` source directory, not the repo root directory. In fact we should probably make a `mnefun/data/` directory if we don't have it yet and put them there

Comment (Eric89GXL) [inline sss_cal.dat]: ...and the `ct_sparse` file, I mean

File: mnefun/_mnefun.py (hunk 8 above)
Comment (kambysese) [inline mnefun/_mnefun.py]: We are for the most part, but I think we should expose these options in case of infant/children data, no?

**Comment by kambysese** [comment]
@Eric89GXL please review. Not sure what conflicts github is complaining about

**Comment by Eric89GXL** [comment]
I'll take a look tomorrow morning. Can you try rebasing to latest `upstream/master`? The "conflicts" refers to a merge conflict that should be solved by rebase.

File: mnefun/_mnefun.py [hunk 12]
```diff
@@ -32,13 +32,14 @@
...
 from mne.label import read_label
 from mne.epochs import combine_event_ids
-try:
-    from mne.chpi import _quat_to_rot, _rot_to_quat
-except ImportError:
-    from mne.io.chpi import _quat_to_rot, _rot_to_quat
+#try:
+#    from mne.chpi import _quat_to_rot, _rot_to_quat
+#except ImportError:
+#    from mne.io.chpi import _quat_to_rot, _rot_to_quat
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: there will be conflicts on these lines, they've been changed in `master`. I didn't do it with a PR but probably should have, sorry about that.

File: mnefun/_mnefun.py [hunk 13]
```diff
@@ -186,6 +187,32 @@ class Params(Frozen):
...
+        identically, choose a buffer length that divides evenly into your data.
+        Any data at the trailing edge that doesn't fit evenly into a whole
+        buffer window will be lumped into the previous buffer.
+    st_correlation : float
+        Correlation limit between inner and outer subspaces used to reject
+        ovwrlapping intersecting inner/outer signals during spatiotemporal SSS.
+    trans_to : str | None
+        The destination location for the head. Can be ``None``, which
+        will not change the head position, or a string path to a FIF file
+        containing a MEG device<->head transformation.
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: Again I'm not sure this is how we want to do this. There were other options that existed already and we should try to unify our treatment of the MaxFilter and `maxwell_filter` operations. Can you try to make it so as many options as possible apply to both versions of the algorithms? We'll need to expand `run_sss.sh` to accept a few more options.

File: mnefun/_mnefun.py [hunk 14]
```diff
@@ -195,11 +222,22 @@ class Params(Frozen):
...
+    - Params has additional properties. Use ``dir(params)`` to see
     all the possible options.
+    - For Maxwell filtering with mne.maxwell_filter the following default
+    parameters:
+        * tSSS correlation = 0.98
+        * Order of internal component of spherical expansion = 8
+        * Order of external component of spherical expansion = 3
+        * tSSS buffer length = 60 seconds
+        * Spherical expansion coordinate frame = head
+        * Coordinate frame origin =  head-digitization-based origin fit
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: For example, we were already using a head-digitization-based origin fit when using MaxFilter. In principle we should be able to do the same things with both (except that eventually mne-python will be more capable, hopefully).

File: mnefun/_mnefun.py [hunk 15]
```diff
@@ -842,6 +897,93 @@ def run_sss_positions(fname_in, fname_out, host='kasga', opts='', port=22):
...
+    -----
+    Compared to Elekta's MaxFilter™ software, Maxwwell filtering in mne-python
+    does not require/support the following arguments:
+        -format
+        -hpicons
+        -autobad
+        -force
+
+    """
+    cal_file = op.join(op.dirname(op.dirname(__file__)), 'mnefun/data/sss_cal.dat')
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: you can get rid of one of the `op.dirname` calls and also get rid of `mnefun`. Also don't use `/` use `op.join` properly

File: mnefun/_mnefun.py [hunk 16]
```diff
@@ -842,6 +897,93 @@ def run_sss_positions(fname_in, fname_out, host='kasga', opts='', port=22):
...
+        raw_files_out = get_raw_fnames(p, subj, 'sss', False, False, run_indices[si])
+        erm_files = get_raw_fnames(p, subj, 'raw', 'only')
+        erm_files_out = get_raw_fnames(p, subj, 'sss', False, False, run_indices[si])
+        prebad_file = _prebad(p, subj)
+        #  process raw files
+        for ii, (r, o) in enumerate(zip(raw_files, raw_files_out)):
+            if not op.isfile(r):
+                raise NameError('File not found (' + r + ')')
+            raw = Raw(r, preload=True, allow_maxshield=True)
+            raw.load_bad_channels(prebad_file, force=True)
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: Need to do `raw.fix_mag_coil_types()` call. Probably also need `mne.chpi.filter_chpi(raw)`, too.

File: mnefun/_mnefun.py [hunk 17]
```diff
@@ -842,6 +897,93 @@ def run_sss_positions(fname_in, fname_out, host='kasga', opts='', port=22):
...
+            else:
+                trans_to = p.trans_to
+
+            # apply maxwell filter
+            raw_sss = maxwell_filter(raw, origin=origin, int_order=p.int_order, ext_order=p.ext_order,
+                                     calibration=cal_file,
+                                     cross_talk=ct_file,
+                                     st_correlation=p.st_correlation, st_duration=st_duration,
+                                     destination=trans_to,
+                                     coord_frame='head')
```
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: need `head_pos` argument

Comment (kambysese) [inline mnefun/_mnefun.py]: Do you want to estimate hp with mne-python or maxfilter?

Comment (Eric89GXL) [inline mnefun/_mnefun.py]: MF for now

**Pull request closed without merging**